### PR TITLE
Use `std::variant` for deferred value

### PR DESCRIPTION
**Pull request by mishasychev**
## Summary
Use `std::variant` to reduce memory consumption for any `lazy` object.

## Checklist
- [x] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [x] I have discussed this with project maintainers:  <br/>
If not checked, I accept that this work might be rejected in favor of a different grand plan.<br/>

## Details
Let's not store the data and initializer separately, because we only need one of them at the moment of time. This is especially important for the _MSVC_ compiler, in which `std::function` has size of 64 bytes on x64 systems.

### CODE REVIEW

File: far/common/lazy.hpp [hunk 1]
```diff
@@ -33,25 +33,27 @@ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+		m_Data(std::move(Initialiser))
 	{
 	}
 
 	T& operator*()
 	{
-		return m_Data?
-			*m_Data :
-			*(m_Data = m_Initialiser());
+		if (initializer_type* compute = std::get_if<initializer_type>(&m_Data))
```
Comment (alabuzhev) [inline far/common/lazy.hpp:53]: Let's use auto, no point in specifying the same type twice on the same line.

File: far/common/lazy.hpp [hunk 2]
```diff
@@ -69,22 +71,15 @@ class lazy
...
 	{
-		m_Data = FWD(rhs);
+		m_Data.template emplace<T>(FWD(rhs));
 		return *this;
 	}
 
 private:
-	std::function<T()> m_Initialiser;
-	std::optional<T> mutable m_Data;
+	std::variant<T, initializer_type> mutable m_Data;
```
Comment (alabuzhev) [inline far/common/lazy.hpp:82]: Neat. Indeed, we only need one.

**Review by alabuzhev — APPROVED** [review]

**Comment by alabuzhev** [comment]
Squash please.

**Comment by mishasychev** [comment]
> Squash please.

Done.

**Comment by alabuzhev** [comment]
Thanks

**Pull request merged**